### PR TITLE
boards: arm: Add support for Silabs EFM32HG-SLSTK3400A

### LINKED ISSUE #6067: Add initial support for Silabs EFM32 Happy Gecko
First add external Gecko SDK for  Silabs EFM32HG family
ext/hal/silabs/gecko/Device/SiliconLabs/EFM32HG/

https://www.silabs.com/products/development-tools/mcu/32-bit/efm32-happy-gecko-starter-kit

### PR DESCRIPTION
**Pull request by chrta**
The EFM32 Happy Gecko USB-enabled kit contains sensors and
peripherals demonstrating some of the EFM32HG MCU capabilities.
This patch add basic support for this board.

Most of the review process was already done in https://github.com/zephyrproject-rtos/zephyr/pull/6945 which was accidentally closed. This fixes #6067

Signed-off-by: Marcio Montenegro <mtuxpe@gmail.com>
Signed-off-by: Christian Taedcke <hacking@taedcke.com>

### CODE REVIEW

File: boards/arm/efm32hg_slstk3400a/doc/efm32hg_slstk3400a.rst [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+| SYSTICK   | on-chip    | systick                             |
++-----------+------------+-------------------------------------+
+| GPIO      | on-chip    | gpio                                |
++-----------+------------+-------------------------------------+
+| USART     | on-chip    | serial port-polling;                |
+|           |            | serial port-interrupt               |
++-----------+------------+-------------------------------------+
+
+The default configuration can be found in the defconfig file:
+
```
Comment (dbkinder) [inline boards/arm/efm32hg_slstk3400a/doc/efm32hg_slstk3400a.rst]: delete the blank line and indent:

    The default configuration can be found in the defconfig file:
    ``boards/arm/efm32hg_slstk3400a/efm32hg_slstk3400a_defconfig``

File: boards/arm/efm32hg_slstk3400a/doc/efm32hg_slstk3400a.rst [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+Build the Zephyr kernel and application:
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/hello_world
+   :board: efm32hg_slstk3400a
+   :goals: build
+
+Connect the EFM32HG-SLSTK3400A to your host computer using the USB port and
+you should see a USB connection that exposes a mass storage device (STK3400)
+and a USB Serial Port. Copy the generated zephyr.bin in the STK3400 drive.
```
Comment (dbkinder) [inline boards/arm/efm32hg_slstk3400a/doc/efm32hg_slstk3400a.rst]: change to:

    generated ``zephyr.bin``

**Review by dbkinder — CHANGES_REQUESTED** [review]
Docs look good .. a couple of recommended tweaks.

**Comment by chrta** [comment]
@dbkinder Both tweaks applied.

**Review by dbkinder — APPROVED** [review]
+1 for docs, thanks!

**Comment by chrta** [comment]
@galak @MaureenHelm Can this be merged now or are there any open issues. Some code review was already done in  #6945

**Comment by carlescufi** [comment]
@galak @MaureenHelm can you take a look?

**Comment by carlescufi** [comment]
The inclusions of `ext/` are zlib-licensed but code under this license from the same vendor already exists in the current tree.

**Comment by carlescufi** [comment]
@chrta can this commit fc51ac2ae96b6e10a159c8339f8ffb5b12739941 be removed and fix it in the build system? Paging @SebastianBoe

**Comment by chrta** [comment]
> CMSIS 5 on the other hand, doesn't seem to include it: https://github.com/ARM-software/CMSIS_5/tree/develop/CMSIS/Core/Include

I think these files got moved to https://github.com/ARM-software/CMSIS_5/tree/develop/CMSIS/DSP/Include

I just checked the Silabs HAL and arm_math.h is only included for the Cortex-M0(+) MCUs, the others do not use this header.

**Comment by chrta** [comment]
It seems that the inclusion of arm_math.h was just some workaround to get the __CLZ() function for a certain compiler. From the old arm_math.h:

~~~
...
/* note: function can be removed when all toolchain support __CLZ for Cortex-M0 */
#if defined (ARM_MATH_CM0_FAMILY) && ((defined (__ICCARM__))  )
  static __INLINE uint32_t __CLZ(
  q31_t data);
...
~~~

I do not think that arm_math.h should be included in this repo, just to support the gecko hal without changes. I would prefer to keep commit https://github.com/zephyrproject-rtos/zephyr/commit/fc51ac2ae96b6e10a159c8339f8ffb5b12739941 in here.

**Comment by carlescufi** [comment]
> I would prefer to keep commit [fc51ac2](https://github.com/zephyrproject-rtos/zephyr/commit/fc51ac2ae96b6e10a159c8339f8ffb5b12739941) in here.

Fine with me. We will need to remember to make this change every time a new version of the HAL is imported but by then hopefully the Gecko HAL will have transitioned to CMSIS 5

**Review by carlescufi — APPROVED** [review]
From the PoV of `ext/` inclusion and board definitions this looks good to me.

File: boards/arm/efm32hg_slstk3400a/board.h [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __INC_BOARD_H
+#define __INC_BOARD_H
+
+#include <soc.h>
+
+/* Push button PB0 */
```
Comment (galak) [inline boards/arm/efm32hg_slstk3400a/board.h]: Can we move the LED & GPIOs in dts,  see PR #10390

**Review by galak — CHANGES_REQUESTED** [review]
Can you look at the additions in PR #10390 related to DTS for GPIO and having LEDs/buttons in DTS.  Thanks

**Comment by chrta** [comment]
Ooops, sorry. I rebased this branch on top of the one from  PR #10390.
When   PR #10390 is merged, i will rebase again on master.

**Review by galak — CHANGES_REQUESTED** [review]
One minor change to move dts_fixup.h into SoC dir and add GPIO fixups

**Review by galak — APPROVED** [review]

**Pull request merged**